pedl_train.py

```diff
@@ -0,0 +1,423 @@
+"""
+This example demonstrates how to port the yolov3 model
+based on PyTorch to PEDL client.
+"""
+import argparse
+import pathlib
+import warnings
+
+from typing import Any, Dict, Sequence, Tuple, Union
+
+# import original yolov3 model files
+import torch
+from torch import nn
+from utils.datasets import LoadImagesAndLabels
+from models import Darknet, attempt_download
+from utils.utils import labels_to_class_weights, compute_loss
+import torch.optim as optim
+from torch.optim.lr_scheduler import _LRScheduler
+
+import pedl
+from pedl.frameworks.pytorch import PyTorchTrial, LRScheduler
+from pedl.frameworks.pytorch.data import DataLoader
+
+import test
+
+from collections import Counter
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+def download_data() -> str:
+    # For a start we hard_code the dataset from bind_mounts
+    # Currently, PEDL Client does not support a download_data_fn
+    return "data"
+
+
+class LazyModule(object):
+    """
+    Since the model definition requires information from the dataloader, we write this class that
+    makes sure it is only initailzed once. This helper class creates a global singleton that lazily
+    initializes a module instance to be used by the trial class and data
+    loaders function.
+    """
+
+    __instance = None
+
+    @staticmethod
+    def get():
+        if LazyModule.__instance is None:
+            # Initializing the augmented dataset on first call
+            train_dataset = LoadImagesAndLabels(
+                train_path,
+                img_size,
+                batch_size,
+                augment=True,
+                hyp=hyp,  # augmentation hyperparameters
+                rect=opt.rect,  # rectangular training
+                cache_labels=True,
+                cache_images=opt.cache_images,
+                single_cls=opt.single_cls,
+            )
+            LazyModule.__instance = train_dataset
```

Comment:
OK - again this was the way it was done originally and basically is a copy-paste of their code that worked "as-is" (and to my own surprise it worked as is). 

Overall it will boil down to the old questions what we want to optimize for:

 1. Minimize the number of code changes when porting an existing model or
 2. Have a good looking code in the end 

For 1, the good thing is that it would be relatively easy to give a complete walkthrough of the steps that were needed to port the model but at the price of ugly final code.

For 2, the good thing is nice looking code but the bad is that it will be harder to understand all the changes that were made and harder to compare to the original code.

Reply:
moved everything in separate `get_config` functions. The only question I have is if it impacts performance if we call this get_config for each batch training. Or if in this case it is better to use global variables.